### PR TITLE
Add RecipesUpdatedEvent

### PR DESCRIPTION
**Pull request by mezz**
Fired when RecipeManager has its recipes updated.

This happens on Client when recipes are synced from the server to the client (just after a client has connected), ~~and on both Server and Client when resource packs are loaded or reloaded.~~

### CODE REVIEW

File: src/main/java/net/minecraftforge/client/event/RecipesUpdatedEvent.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+public class RecipesUpdatedEvent extends Event
+{
+    private final Collection<IRecipe> recipes;
+
+    public RecipesUpdatedEvent(Collection<IRecipe> recipes)
+    {
+        this.recipes = Collections.unmodifiableCollection(recipes);
+    }
+
+    public Collection<IRecipe> getRecipes()
```
Comment (williewillus) [inline src/main/java/net/minecraftforge/client/event/RecipesUpdatedEvent.java]: maybe some docs on what exactly this holds? All the recipes or the ones that changed/reloaded?

Comment (mezz) [reply src/main/java/net/minecraftforge/client/event/RecipesUpdatedEvent.java]: It's always the same as `RecipeManager#getRecipes` so I'm actually thinking of removing this from the event.
Thoughts?

Comment (Landmaster) [reply src/main/java/net/minecraftforge/client/event/RecipesUpdatedEvent.java]: As long as that fact is documented in `RecipesUpdatedEvent`, that should be fine.

Comment (mezz) [reply src/main/java/net/minecraftforge/client/event/RecipesUpdatedEvent.java]: Ok I have updated the event so it does not include the recipes. They are accessible directly from the recipe manager.

File: src/main/java/net/minecraftforge/client/event/RecipesUpdatedEvent.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+import net.minecraft.item.crafting.RecipeManager;
+import net.minecraftforge.api.distmarker.Dist;
+import net.minecraftforge.eventbus.api.Event;
+
+/**
+ * Fired when {@link RecipeManager} has all of its recipes updated.
+ *
+ * This happens on {@link Dist#CLIENT} when recipes are synced from the server to the client (just after a client has connected),
+ * and on both {@link Dist#DEDICATED_SERVER} and {@link Dist#CLIENT} when resource packs are loaded or reloaded.
```
Comment (williewillus) [inline src/main/java/net/minecraftforge/client/event/RecipesUpdatedEvent.java]: should this say datapack? recipes aren't stored in resource packs anymore

**Comment by Shadows-of-Fire** [comment]
It may be better to commit this event to just recipe sync, as reloading could be left to a reload listener when selective server reloading is implemented.

**Comment by mezz** [comment]
There is no ordering to the resource listeners so I wanted something explicit.
I'll take a look at when the RecipeManager adds its resource listener to see if there is a reliable place to register after it though, if there is then we don't need it as part of this event.

**Comment by mezz** [comment]
I have updated this to only fire the event when the recipes are synced from the server.
A selective reload listener should be good for the rest.

**Pull request merged**